File: saewoo1/day5/19.java
/* 뒤에서부터 n번째에 위치한 노드를 제거하세요 */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size = 0;
        for (ListNode node = head; node != null; node = node.next) // 노드 개수 세기
            size++;
        if (size == 1)
            return null;
        if (size == n)
            return head.next; // 여기까지 예외처리
        ListNode res = head;
        for (int i = 0; i < size - 1 - n; i++) { //뒤에서 n번째에 위치한 노드 전까지 위치하도록
            res = res.next;
        }
        res.next = res.next.next; // 제거할 노드를 그 다음 노드로 덮어 씌웁니다
        return head; // pointer로 이어졌으니 next로 이동하지 않은 첫 노드를 가리키는 head를 반환하도록
    }
}

File: oio337a/Algorithm_1/day_5/19.py
# 순회해서 길이를 구하고 n 만큼 적게 순환 돈뒤 다음 노드와 이어주기

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        end = start = head
        length = 1

        while end.next: # node 길이 재버리기
            end = end.next
            length += 1
        if length == 1 and n == 1: # 예외 처리 하나만있을 때
            head.val = ''
            return head
        if length - n == 0: # 예외 처리 제거할게 없을 때
            return head.next
        for _ in range(length - n - 1): # 순회하며 원하는 위치 찾기
            start = start.next
        start.next = start.next.next # 값 스킵하기
        return head
        